Added E2E PermuteOp support

azecevicTT opened the pull request:
Added support for PermuteOp, since it's already supported in [ttnn](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.permute.html#ttnn.permute).

`stablehlo.transpose` and `ttir.permute` have the same semantics, so decomposition of `stablehlo.transpose` into series of `ttir.transpose`es is not needed anymore.

Closes https://github.com/tenstorrent/tt-mlir/issues/652

mtopalovicTT commented:
@azecevicTT can we check this in? I need it for a fix. I don't see anything blocking here.

sdjordjevicTT commented on include/ttmlir/Dialect/TTIR/IR/TTIROps.td:
```diff
@@ -1168,6 +1168,34 @@ def TTIR_MatmulOp : TTIR_DPSOp<"matmul"> {
+    let arguments = (ins AnyRankedTensor:$input,
+                         AnyRankedTensor:$output,
+                         DenseI64ArrayAttr:$permutation,
+                         TT_OperandConstraintArrayAttr:$operand_constraints);
```
Don't forget to remove operand_constraints :)

sdjordjevicTT commented on runtime/lib/ttnn/operations/data_movement/permute.cpp:
```diff
@@ -0,0 +1,31 @@
+  std::vector<int64_t> permutation(op->permutation()->begin(),
+                                   op->permutation()->end());
+
+  // Has to called with the verbose invoke, since calling it with composite=true
```
Do we need to create an issue on the Metal side regarding this?

azecevicTT replied:
Yeah, I created an issue last week and it seems they have a fix that will land on the main soon https://github.com/tenstorrent/tt-metal/pull/15881. That's why I moved this PR to draft, to not cause an issue during uplift.

mtopalovicTT commented on include/ttmlir/Utils.h:
```diff
@@ -72,17 +72,11 @@ constexpr std::underlying_type_t<Enum> enum_as_int(Enum e) {
-    }
-  }
-  return result.str();
+template <typename Range>
```
nit:  just some comment what this does and returns

mtopalovicTT commented on lib/Dialect/TTIR/IR/TTIROps.cpp:
```diff
@@ -1385,6 +1386,44 @@ ::mlir::LogicalResult mlir::tt::ttir::ScatterOp::verify() {
+      !std::is_permutation(permutation.begin(), permutation.end(),
+                           dimensions.begin())) {
+    return emitOpError("Expected a permutation of {k | 0 <= k < " +
+                       std::to_string(inputRank) + "} got (" +
```
not sure if it make sense for better error but maybe:
```
"Expected a permutation of (" +
ttmlir::utils::join(dimensions)" ...
```

mtopalovicTT commented on test/ttmlir/Dialect/TTIR/permute/permute_tests_negative.mlir:
```diff
@@ -0,0 +1,36 @@
+// RUN: not ttmlir-opt --split-input-file %s 2>&1 | FileCheck %s
+// Negative tests for matmul operation
```
nit: fix comment

svuckovicTT commented on include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:
```diff
@@ -915,4 +915,25 @@ def TTNN_ReduceScatterOp: TTNN_Op<"reduce_scatter"> {
+      %0 = "ttir.permute"(%a) {permutation = array<i64: 1, 2, 0>} : (tensor<2x3x4xi32>) -> tensor<3x4x2xi32>
+    }];
+
+    let arguments = (ins AnyRankedTensor:$input,
+                         DenseI64ArrayAttr:$permutation);
```
These are the signatures I'm seeing in metal (unless they've been changed recently):
```c
struct ExecutePermute {
    static ttnn::Tensor invoke(
        uint8_t queue_id,
        const ttnn::Tensor& input_tensor,
        tt::stl::Span<const int64_t> dims,
        const std::optional<MemoryConfig>& memory_config,
        bool composite = true,
        const std::optional<float>& pad_value = 0.0f);

    static ttnn::Tensor invoke(
        const ttnn::Tensor& input_tensor,
        tt::stl::Span<const int64_t> dims,
        const std::optional<MemoryConfig>& memory_config,
        const std::optional<float>& pad_value = 0.0f);

    static ttnn::Tensor invoke(
        const ttnn::Tensor& input_tensor,
        tt::stl::Span<const int64_t> dims,
        const std::optional<float>& pad_value = 0.0f);
};
```

First one doesn't work for us due to queue_id, but I'd suggest we pick one of the remaining two, and implement all the operands that are there.

azecevicTT replied:
I went with the second one.

svuckovicTT commented on runtime/lib/ttnn/operations/data_movement/permute.cpp:
```diff
@@ -0,0 +1,31 @@
+  const ::ttnn::Tensor &in = tensorPool.at(op->in()->global_id());
+  DEBUG_ASSERT(in.is_allocated());
+
+  std::vector<int64_t> permutation(op->permutation()->begin(),
```
Consider using smallvector

azecevicTT replied:
I deliberately went with `std::vector` in the runtime, as the runtime is totally independent of the LLVM. In my opinion, runtime should be the only exception from "prefer the LLVM ADT to std structures" in our software stack.

svuckovicTT commented on test/ttmlir/Silicon/StableHLO/Unary/tranpose_op.mlir:
```diff
@@ -9,8 +9,8 @@
     // CHECK-LABEL: func.func public @test_transpose
-    // CHECK: ttnn.transpose
-    // CHECK-SAME: {dim0 = 1 : si32, dim1 = 0 : si32}
+    // CHECK: "ttnn.permute"
+    // CHECK-SAME: permutation = array<i64: 1, 0>
```
Does this mean that we'll effectively never have a `ttnn.transpose` in the IR?

azecevicTT commented:
FYI @mtopalovicTT @sdjordjevicTT @svuckovicTT I've made non-trivial changes since you've reviewed this PR. I've refactored the TTIRToTTIR decomposition code, where I replaced series of `ttir.transpose`s to one `ttir.permute`. It turned out to be a bigger change than what I expected, I should've probably split it into a separate PR, so I'm sorry for the inconvenience.
@LPanosTT Can you also review conv decomposition changes?

mtopalovicTT commented:
@azecevicTT will do today tnx

LPanosTT reviewed (approved):
Looks good to me, thanks for patching up the Conv and pooling decomps!

svuckovicTT commented on include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:
```diff
@@ -1023,4 +1023,27 @@ def TTNN_MeshShardOp: TTNN_Op<"mesh_shard"> {
+    }];
+
+    let arguments = (ins AnyRankedTensor:$input,
+                         DenseI64ArrayAttr:$permutation,
```
I'm somewhat torn on what to do here... In the lib, there's a `Span` object, which seems to inherit from boost's span, which seems to be yet another container for objects. I'd like to differentiate it from a `DenseI64ArrayAttr` so that the RT doesn't need to "know" about this (the overloaded methods for processing operands talk...).

azecevicTT replied:
I don't see a problem, `DenseI64ArrayAttr` is MLIRs equivalent of `std::vector`, `span` on the other hand is just a view over array-like object, similar to `ArrayRef` in LLVM, so if we always convert `DenseI64ArrayAttr` to `std::vector` it's going to work out of the box.

svuckovicTT replied:
I did not know that, that's wonderful, thank you! 😊

jnie-TT commented on runtime/lib/ttnn/operations/data_movement/permute.cpp:
```diff
@@ -0,0 +1,30 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "permute.h"
```
Nit: `operations/data_movement/permute.h`

jnie-TT reviewed (approved):
Minor nit inline, otherwise runtime changes look good

tapspatel reviewed (approved):
looks good! thanks for adding the relevant test cases
